### PR TITLE
sourcery 0.5.3 (new formula)

### PR DESCRIPTION
**Pull request by vknabel**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

[Sourcery](https://github.com/krzysztofzablocki/Sourcery) scans your source code, applies your personal templates and generates Swift code for you, allowing you to use meta-programming techniques to save time and decrease potential mistakes.

### CODE REVIEW

**Comment by woodruffw** [comment]
@BrewTestBot test this please

File: Formula/sourcery.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+class Sourcery < Formula
+  desc "Meta-programming for Swift, stop writing boilerplate code."
+  homepage "https://github.com/krzysztofzablocki/Sourcery"
+  url "https://github.com/krzysztofzablocki/Sourcery/archive/0.5.3.tar.gz"
+  sha256 "ad34e2146e73ce81894ac1dd9130fb2fd12c8213789aef7c5cde9580f813c823"
+
+  depends_on :xcode => ["8.0", :build]
+
+  def install
+    system "unset CC; swift build -c release"
```
Comment (woodruffw) [inline Formula/sourcery.rb]: Why is unsetting `CC` necessary?

File: Formula/sourcery.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    bin.install ".build/release/sourcery"
+    lib.install Dir[".build/release/*.dylib"]
+  end
+
+  test do
+    system "#{bin}/sourcery", "--version"
```
Comment (woodruffw) [inline Formula/sourcery.rb]: Any chance of a more substantial test?

File: Formula/sourcery.rb (hunk 1 above)
Comment (vknabel) [inline Formula/sourcery.rb]: `swift build -c release` failed with `swift-build: error: invalid inferred toolchain: could not find `clang` at expected path /private/tmp/sourcery-20170125-20504-188t29n/Sourcery-0.5.3/clang`

Is there a different workaround? (Running `xcrun swift build -c release` failed, too)

**Comment by vknabel** [comment]
@woodruffw I added some tests, but they are failing because of two reasons (where both work outside `brew test sourcery`):

1. `sourcery` cannot create `~/Library/Caches/Sourcery`, which leads to the error in Jenkins (operation not permitted). When creating `~/Library/Caches/Sourcery` manually, the next error 2 will come in. When executing `sourcery` outside `brew test sourcery` it will create its caches directory itself.

2. Running `sourcery` with the test params fails with an `zsh: illegal hardware instruction  /usr/local/Cellar/sourcery/0.5.3/bin/sourcery`. 

```zsh
# in brew test sourcery --debug
% /usr/local/Cellar/sourcery/0.5.3/bin/sourcery /tmp/sourcery-test-20170125-2654-k768xq/Test.swift /tmp/sourcery-test-20170125-2654-k768xq/Test.stencil /tmp/sourcery-test-20170125-2654-k768xq/Generated.swift
Scanning sources...
zsh: illegal hardware instruction  /usr/local/Cellar/sourcery/0.5.3/bin/sourcery
```

When using my own shell, everything works as expected when using the same files and binaries.

```zsh
# in my own shell
-> /usr/local/Cellar/sourcery/0.5.3/bin/sourcery /tmp/sourcery-test-20170125-2654-k768xq/Test.swift /tmp/sourcery-test-20170125-2654-k768xq/Test.stencil /tmp/sourcery-test-20170125-2654-k768xq/Generated.swift
Scanning sources...
Loading templates...
Loaded 1 templates.
Generating code...
Processing time 0.141945004463196 seconds
```

Do you have got an idea, what may explain this kind of behavior?

File: Formula/sourcery.rb (hunk 1 above)
Comment (woodruffw) [inline Formula/sourcery.rb]: Put `ENV.delete("CC")` on the line above.

**Comment by woodruffw** [comment]
Formulae cannot write outside of the sandbox during the `install` or `test` phases, so `sourcery` needs to be told to cache somewhere other than `~/Library/Caches`. Can it take an environment variable or flag to that effect?

> zsh: illegal hardware instruction /usr/local/Cellar/sourcery/0.5.3/bin/sourcery

If I remember correctly, this is caused by targeting a CPU with a newer instruction set than the host. I'm not familiar with `swift build`, but can it be told to target the host instead?

File: Formula/sourcery.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  def install
+    system "unset CC; swift build -c release"
+    bin.install ".build/release/sourcery"
+    lib.install Dir[".build/release/*.dylib"]
+  end
+
+  test do
+    assert_match version.to_s, shell_output("#{bin}/sourcery --version").chomp
+    (testpath/"Test.swift").write "enum One { }\nenum Two { }"
+    (testpath/"Test.stencil").write "// Found {{ types.all.count }} Types\n// {% for type in types.all %}{{ type.name }}, {% endfor %}"
```
Comment (woodruffw) [inline Formula/sourcery.rb]: Please use a heredoc here.

File: Formula/sourcery.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    system "#{bin}/sourcery", testpath/"Test.swift", testpath/"Test.stencil", testpath/"Generated.swift"
+    assert_equal "// Generated using Sourcery 0.5.3 — https://github.com/krzysztofzablocki/Sourcery
```
Comment (woodruffw) [inline Formula/sourcery.rb]: Here as well.

File: Formula/sourcery.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  depends_on :xcode => ["8.0", :build]
+
+  def install
+    system "unset CC; swift build -c release"
+    bin.install ".build/release/sourcery"
+    lib.install Dir[".build/release/*.dylib"]
+  end
+
+  test do
+    assert_match version.to_s, shell_output("#{bin}/sourcery --version").chomp
```
Comment (woodruffw) [inline Formula/sourcery.rb]: You can remove this test if we're going to use the newly added one.

File: Formula/sourcery.rb (hunk 4 above)
Comment (woodruffw) [inline Formula/sourcery.rb]: (And prefer `assert_match` so that you don't have to call `strip` after reading)

**Comment by vknabel** [comment]
@woodruffw prefixing `swift build` with `xcrun -sdk macosx` fixed the compilation issue for me.

Regarding the Sandbox issue: currently I kept the 'version assertion' in. Shall I comment out the real test and add a comment pointing to an issue (I will create) on the Sourcery repo?

Something like this:
```ruby
# Tests are temporarily disabled because of sandbox issues, 
# as Sourcery tries to write to ~/Library/Caches/Sourcery
# See https://github.com/krzysztofzablocki/Sourcery/issues/XXX
#
# Remove this test once the issue has been closed
assert_match version.to_s, shell_output("#{bin}/sourcery --version").chomp

# Uncomment these tests when the issue has been closed
# (testpath/"Test.swift").write <<-TEST_SWIFT

**Comment by woodruffw** [comment]
> prefixing swift build with xcrun -sdk macosx fixed the compilation issue for me.

Good to hear. `xcrun -sdk macosx` really just updates the `SDKROOT` environment variable,
so this should work and saves us an invocation:

```ruby
ENV.delete("CC")
ENV["SDKROOT"] = MacOS.sdk_path
system "xcrun -sdk macosx swift build -c release"
```

> Regarding the Sandbox issue: currently I kept the 'version assertion' in. Shall I comment out the real test and add a comment pointing to an issue (I will create) on the Sourcery repo?

Hmm. If the new tests will require an upstream change to fix, let's just remove them for now and go with only the version assertion. Then, when upstream applies the changes and releases a new version, we can re-add the more detailed tests.

**Comment by vknabel** [comment]
@woodruffw I commented out all tests except the version test.

Additionally I fixed a `brew audit sourcery` issue.

**Comment by swizzlr** [comment]
@woodruffw @vknabel Anything I can do to help?

**Comment by woodruffw** [comment]
@vknabel  @swizzlr Nope, this looks good to me; I've just been a bit busy. I'll :ship: it tonight, unless someone else gets to it first :smile:

**Comment by swizzlr** [comment]
@woodruffw no worries, thanks dude – and thanks @vknabel for pushing this forward!

**Comment by vknabel** [comment]
@woodruffw @swizzlr thanks for your time, help and ideas!

**Comment by woodruffw** [comment]
Thanks @vknabel, and congratulations on your first Homebrew contribution/formula! :tada:

**Comment by ilovezfs** [comment]
I'm a bit confused as it looks like the referenced PR https://github.com/krzysztofzablocki/Sourcery/pull/133 was merged but the sandbox issue persists.

**Comment by ilovezfs** [comment]
https://github.com/Homebrew/homebrew-core/pull/11407

**Comment by vknabel** [comment]
@ilovezfs I am sorry for your confusion.
I already wanted to update the tests appropriately, but I currently haven't got enough time.
According to the source code changes, one needs to add `--disableCache` to the invocation of `sourcery`. 
I hope that fixes the sandbox error.

**Comment by ilovezfs** [comment]
@vknabel strangely that did not seem to help:
```
==> /usr/local/Cellar/sourcery/0.5.9/bin/sourcery --disableCache --sources /tmp/sourcery-test-20170321-24134-1h3z620/Test.swift --templates /tmp/sourcery-test-20170321-24134-1h3z620/Test.stencil
/usr/local/Homebrew/Library/Homebrew/debrew.rb:11:in `raise'
BuildError: Failed executing: /usr/local/Cellar/sourcery/0.5.9/bin/sourcery --disableCache --sources /tmp/sourcery-test-20170321-24134-1h3z620/Test.swift --templates /tmp/sourcery-test-20170321-24134-1h3z620/Test.stencil
1. raise
2. ignore
3. backtrace
4. irb
5. shell
Choose an action: 5
When you exit this shell, you will return to the menu.
bash-3.2$ /usr/local/Cellar/sourcery/0.5.9/bin/sourcery --disableCache --sources /tmp/sourcery-test-20170321-24134-1h3z620/Test.swift --templates /tmp/sourcery-test-20170321-24134-1h3z620/Test.stencil
Scanning sources...
Illegal instruction: 4
bash-3.2$ exit
==> Kept temporary files
Temporary files retained at /tmp/sourcery-test-20170321-24134-1h3z620
Error: sourcery: failed
undefined class/module Debrew::
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:39:in `load'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:39:in `block (3 levels) in safe_fork'
/usr/local/Homebrew/Library/Homebrew/utils.rb:380:in `ignore_interrupts'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:26:in `block (2 levels) in safe_fork'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:7:in `open'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:7:in `block in safe_fork'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/tmpdir.rb:88:in `mktmpdir'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:6:in `safe_fork'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/test.rb:64:in `block in test'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/test.rb:29:in `each'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/test.rb:29:in `test'
/usr/local/Homebrew/Library/Homebrew/brew.rb:91:in `<main>'
iMac-TMP:homebrew-core joe$ brew test sourcery -vd --no-sandbox
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/sourcery.rb
Testing sourcery
/usr/local/Homebrew/Library/Homebrew/test.rb (Formulary::FromPathLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/sourcery.rb
==> /usr/local/Cellar/sourcery/0.5.9/bin/sourcery --disableCache --sources /tmp/sourcery-test-20170321-24289-171lnnd/Test.swift --templates /tmp/sourcery-test-20170321-24289-171lnnd/Test.stencil
Scanning sources...
Found 2 types.
Loading templates...
Loaded 1 templates.
Generating code...
Finished.
Processing time 0.0180209875106812 seconds
```

```

  test do
    (testpath/"Test.swift").write <<-TEST_SWIFT
    enum One { }
    enum Two { }
    TEST_SWIFT

    (testpath/"Test.stencil").write <<-TEST_STENCIL
    // Found {{ types.all.count }} Types
    // {% for type in types.all %}{{ type.name }}, {% endfor %}
    TEST_STENCIL

    system "#{bin}/sourcery", "--disableCache", "--sources", testpath/"Test.swift", "--templates", testpath/"Test.stencil"

    expected = <<-GENERATED_SWIFT.undent
      // Generated using Sourcery 0.5.9 — https://github.com/krzysztofzablocki/Sourcery
      // DO NOT EDIT

          // Found 2 Types
          // One, Two,
    GENERATED_SWIFT
    assert_match expected, (testpath/"Test.generated.swift").read, "sourcery generation failed"
  end
```

**Comment by Antondomashnev** [comment]
I've been trying to narrow it down and what I can see is that the test below works:
```
assert_match version.to_s, shell_output("#{bin}/sourcery --version").chomp

    test_swift = testpath/"Test.swift"
    test_stencil = testpath/"Test.stencil"
    test_output = testpath/"Generated.swift"

    test_swift.write <<-TEST_SWIFT
enum One { }
enum Two { }
    TEST_SWIFT

    test_stencil.write <<-TEST_STENCIL
Found {{ types.all.count }} Types
{% for type in types.all %}{{ type.name }}, {% endfor %}
    TEST_STENCIL

    expected = "// Generated using Sourcery #{version.to_s} — https://github.com/krzysztofzablocki/Sourcery\n// DO NOT EDIT\n\n\nFound 2 Types\nOne, Two,"

    system "#{bin}/sourcery", "--sources", "/Users/antondomashnev/Desktop/Temporary/Test.swift", "--templates", "/Users/antondomashnev/Desktop/Temporary/Test.stencil", "--output", "#{test_output}"

    assert_match expected, (test_output).read, "sourcery generation failed"
```

Where the important part is the path to `--sources` and `--templates`. But similar test with the paths to `--sources` and `--templates` in `/tmp` dir doesn't work:
```
assert_match version.to_s, shell_output("#{bin}/sourcery --version").chomp

    test_swift = testpath/"Test.swift"
    test_stencil = testpath/"Test.stencil"
    test_output = testpath/"Generated.swift"

    test_swift.write <<-TEST_SWIFT
enum One { }
enum Two { }
    TEST_SWIFT

    test_stencil.write <<-TEST_STENCIL
Found {{ types.all.count }} Types
{% for type in types.all %}{{ type.name }}, {% endfor %}
    TEST_STENCIL

    expected = "// Generated using Sourcery #{version.to_s} — https://github.com/krzysztofzablocki/Sourcery\n// DO NOT EDIT\n\n\nFound 2 Types\nOne, Two,"

    system "#{bin}/sourcery", "--sources", "#{test_swift}", "--templates", "#{test_stencil}", "--output", "#{test_output}"

    assert_match expected, (test_output).read, "sourcery generation failed"
```

**Comment by krzysztofzablocki** [comment]
@Antondomashnev try with `--disableCache` ?

**Comment by Antondomashnev** [comment]
@krzysztofzablocki I've tried already, didn't help.

**Pull request closed without merging**